scipy/stats/_mannwhitneyu.py

```diff
@@ -0,0 +1,430 @@
+    x = np.moveaxis(x, axis, -1)
+    y = np.moveaxis(y, axis, -1)
+    z = np.broadcast(x[..., 0], y[..., 0])
+    x = np.broadcast_to(x, z.shape + (x.shape[-1],))
+    y = np.broadcast_to(y, z.shape + (y.shape[-1],))
+    z = np.concatenate((x, y), axis=-1)
+    return x, y, z
+
+
+class _MWU:
+    '''Distribution of MWU statistic under the null hypothesis'''
+    # Possible improvement: if m and n are small enough, use integer arithmetic
+
+    def __init__(self):
+        '''Minimal initializer'''
+        self._fmnks = -np.ones((1, 1, 1))
+
+    def pmf(self, k, m, n):
+        '''Probability mass function'''
+        self._resize_fmnks(m, n, np.max(k))
+        # could loop over just the unique elements, but probably not worth
+        # the time to find them
+        for i in np.ravel(k):
+            self._f(m, n, i)
+        return self._fmnks[m, n, k] / special.binom(m + n, m)
+
+    def cdf(self, k, m, n):
+        '''Cumulative distribution function'''
+        # We could use the fact that the distribution is symmetric to avoid
+        # summing more than m*n/2 terms, but it might not be worth the
+        # overhead. Let's leave that to an improvement.
+        pmfs = self.pmf(np.arange(0, np.max(k) + 1), m, n)
+        cdfs = np.cumsum(pmfs)
+        return cdfs[k]
+
+    def sf(self, k, m, n):
+        '''Survival function'''
+        # Use the fact that the distribution is symmetric; i.e.
+        # _f(m, n, m*n-k) = _f(m, n, k), and sum from the left
+        k = m*n - k
+        # Note that both CDF and SF include the PMF at k. The p-value is
+        # calculated from the SF and should include the mass at k, so this
+        # is desirable
+        return self.cdf(k, m, n)
+
+    def _resize_fmnks(self, m, n, k):
+        '''If necessary, expand the array that remembers PMF values'''
+        # could probably use `np.pad` but I'm not sure it would save code
+        shape_old = np.array(self._fmnks.shape)
+        shape_new = np.array((m+1, n+1, k+1))
+        if np.any(shape_new > shape_old):
+            shape = np.maximum(shape_old, shape_new)
+            fmnks = -np.ones(shape)             # create the new array
+            m0, n0, k0 = shape_old
+            fmnks[:m0, :n0, :k0] = self._fmnks  # copy remembered values
+            self._fmnks = fmnks
+
+    def _f(self, m, n, k):
+        '''Recursive implementation of function of [3] Theorem 2.5'''
+
+        fmnks = self._fmnks  # for convenience
+
+        # [3] Theorem 2.5 Line 1
+        if k < 0 or m < 0 or n < 0 or k > m*n:
+            return 0
+
+        # if already calculated, return the value
+        if fmnks[m, n, k] >= 0:
+            return fmnks[m, n, k]
+
+        if k == 0 and m >= 0 and n >= 0:  # [3] Theorem 2.5 Line 2
+            fmnk = 1
+        else:   # [3] Theorem 2.5 Line 3 / Equation 3
+            fmnk = self._f(m-1, n, k-n) + self._f(m, n-1, k)
+
+        fmnks[m, n, k] = fmnk  # remember result
```

Comment:
Does this update self._fmnks? Maybe easier to read if you write self._fmnks[...] = ...

Reply:
Yes because above, `fmnks = self._fmnks  # for convenience`. But feel free to commit the suggested change if you prefer.

Reply:
ok, since one cannot directly see the type of the variables and if they are mutable, I personally prefer the changed version

Reply:
OK. This is resolved, right?

Reply:
yes